query-algebrizer/src/clauses/mod.rs

```diff
@@ -825,7 +826,139 @@ impl ConjoiningClauses {
     /// where `?y` must simultaneously be a ref-typed attribute and a boolean-typed attribute. This
     /// function deduces that and calls `self.mark_known_empty`. #293.
     pub fn expand_type_tags(&mut self) {
-        // TODO.
+        for (v, tag) in self.extracted_types.iter() {
+            let cols = self.column_bindings.get(v).unwrap().clone();
+            for col in cols {
+                if col.0 != tag.0 { 
+                    let value = QueryValue::Column(QualifiedAlias(col.0.clone(), Column::Fixed(DatomsColumn::ValueTypeTag)));
+                    let alias = tag.clone();
+                    self.wheres.0.push(ColumnConstraintOrAlternation::Constraint(ColumnConstraint::Equals(alias, value)));
+                }
+            }
+        }
+
+        let mut computed_types: BTreeMap<Variable, TypedValue> = BTreeMap::default();
+
+        for computed_table in self.computed_tables.iter() {
+            match computed_table {
+                &ComputedTable::NamedValues { ref names, ref values } => {
+                    for (i, item) in names.iter().enumerate() {
+                        computed_types.insert(item.clone(), values[i].clone());
+                    }
+                },
+                _ => (),
+            }
+        }
+
+        let mut empty_reason: Option<EmptyBecause> = None;
+        let mut wheres_to_add = vec![];
+        for (v, value_type_set) in self.known_types.iter() {
```

Comment:
Any variable for which `value_type_set.is_unit()`, or any variable for which all types in the `value_type_set` map to a different SQLite storage type (so strings and doubles are safe to compare, but doubles = longs, and instants = longs), is safe: simply comparing the value is adequate, right?